modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java

```diff
@@ -0,0 +1,168 @@
+ */
+public class RebalanceRaftGroupEventsListener implements RaftGroupEventsListener {
+
+    /** Meta storage manager. */
+    private final MetaStorageManager metaStorageMgr;
+
+    /** Table configuration instance. */
+    private final TableConfiguration tblConfiguration;
+
+    /** Unique partition id. */
+    private final String partId;
+
+    /** Partition number. */
+    private final int partNum;
+
+    /** Busy lock of parent component for synchronous stop. */
+    private IgniteSpinBusyLock busyLock;
+
+    /** Resolver that resolves a network address to cluster node. */
+    private final Function<NetworkAddress, ClusterNode> clusterNodeRslvr;
+
+    /**
+     * Constructs new listener.
+     *
+     * @param metaStorageMgr Meta storage manager.
+     * @param tblConfiguration Table configuration.
+     * @param partId Partition id.
+     * @param partNum Partition number.
+     */
+    public RebalanceRaftGroupEventsListener(
+            MetaStorageManager metaStorageMgr,
+            TableConfiguration tblConfiguration,
+            String partId,
+            int partNum,
+            IgniteSpinBusyLock busyLock,
+            Function<NetworkAddress, ClusterNode> clusterNodeRslvr) {
+        this.metaStorageMgr = metaStorageMgr;
+        this.tblConfiguration = tblConfiguration;
+        this.partId = partId;
+        this.partNum = partNum;
+        this.busyLock = busyLock;
+        this.clusterNodeRslvr = clusterNodeRslvr;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onLeaderElected() {
+        // TODO: IGNITE-16800 implement this method
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onNewPeersConfigurationApplied(List<PeerId> peers) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteInternalException(new NodeStoppingException());
+        }
+
+        try {
+            doOnNewPeersConfigurationApplied(peers);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void onReconfigurationError(Status status) {
+        // TODO: IGNITE-14873 implement this method
+    }
+
+    /**
+     * Implementation of {@link RebalanceRaftGroupEventsListener#onNewPeersConfigurationApplied(List)}.
+     *
+     * @param peers Peers
+     */
+    private void doOnNewPeersConfigurationApplied(List<PeerId> peers) {
+        Map<ByteArray, Entry> keys = metaStorageMgr.getAll(
+                Set.of(partAssignmentsPlannedKey(partId), partAssignmentsPendingKey(partId))).join();
+
+        Entry plannedEntry = keys.get(partAssignmentsPlannedKey(partId));
+
+        List<ClusterNode> appliedPeers = peers
+                .stream()
+                .map(p -> clusterNodeRslvr.apply(
+                        NetworkAddress.from(p.getEndpoint().getIp() + ":" + p.getEndpoint().getPort())))
+                .collect(Collectors.toList());
+
+        tblConfiguration.change(ch -> {
```

Comment:
Let's add a test, **within separate ticket**, that will check that everything works fine, if node files after tblConfiguration.change but before ms.invoke. I expect that it should work properly, but nevertheless the test will be useful.

Reply:
Thanks, added this case to https://issues.apache.org/jira/browse/IGNITE-14873 in comments